cpp/src/arrow/util/hashing.h

```diff
@@ -143,6 +144,21 @@ struct ScalarHelper<Scalar, AlgNum, enable_if_t<std::is_floating_point<Scalar>::
   }
 };
 
+template <typename Scalar, uint64_t AlgNum>
+struct ScalarHelper<Scalar, AlgNum,
+                    enable_if_t<std::is_same_v<Scalar, ::arrow::util::Float16>>>
+    : public ScalarHelperBase<Scalar, AlgNum> {
+  // ScalarHelper specialization for Float16
+
+  static bool CompareScalars(Scalar u, Scalar v) {
+    if (u.is_nan()) {
+      // XXX should we do a bit-precise comparison?
```

Comment:
I think we should consider any two NaNs equivalent. Bit-precise comparison can be accomplished by viewing as `uint16_t`

Reply:
FWIW: I meant to say that I think the current behavior of considering all NaNs equivalent is correct and the comment should be removed.